### PR TITLE
[TRTLLM-14833][chore] Introduce executor/ray/ for Ray executor integration

### PR DESCRIPTION
**Pull request by lori-ren**
> [!IMPORTANT]
> **Please do not squash-merge this PR.**
>
> The first commit is a pure `git mv`; the second carries everything else. Squashing collapses
> them into one commit where the old paths are *modified* — they hold the compatibility shims —
> rather than *deleted*, so no rename is detected and `git blame` reattributes every line of the
> moved files to whoever merges. A merge or rebase merge keeps the history intact.
>
> Both commit titles already carry the `(#17487)` suffix that a non-squash merge would otherwise
> not add.

## Dev Engineer Review

- Consolidates Ray integration under `tensorrt_llm.executor.ray`.
- Adds `RayExecutor`, `RayGPUWorker`, `RayWorkerWrapper`, Ray utilities, and the Ray fallback stub.
- Preserves legacy imports through compatibility shims with deprecation warnings.
- Updates internal imports, documentation links, Ruff configuration, pre-commit configuration, and legacy file lists.
- Adds validation for Ray worker setup, placement groups, distributed execution, request handling, and shutdown.
- Configuration changes match the new file layout. No unintended scope changes are indicated.
- The compatibility design avoids requiring Ray for stub-only imports and defers runtime failures until Ray functionality is used.
- CI results remain inconclusive. Multiple `/bot run` executions failed, including the associated `L0_MergeRequest_PR` pipeline. The latest run has no completion result.

## QA Engineer Review

Modified CI test lists:

- `tests/integration/test_lists/test-db/l0_cpu.yml`: added `unittest/executor/test_ray_stub.py`.
- `tests/integration/test_lists/test-db/l0_h100.yml`: added `unittest/executor/test_shim_ray.py`.

Added test functions:

- `test_ray_stub.py` covers fallback imports, bare and parameterized `remote` decorators, missing attributes, installation guidance, and distributed fallback behavior.
- `test_shim_ray.py` verifies forwarded object identity, defining module names, and one warning with the replacement path.

Both test files are covered by `test-db/` CI lists. Verdict: sufficient.

## Description

Group all Ray integration under one executor sub-package:
- _ray_utils.py to executor/ray/utils.py
- ray_stub.py to executor/ray/stub.py
- executor/ray_executor.py to executor/ray/executor.py
- executor/ray_gpu_worker.py to executor/ray/gpu_worker.py

## Test Coverage

- shim layer for replaced source files
- tests/unittest/_torch/ray_orchestrator import pass

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: tensorrt_llm/executor/ray/stub.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        return decorator(args[0])
+
+    return decorator
+
+
+def __getattr__(name):
+    msg = f'Ray not installed, so "ray.{name}" is unavailable.'
+    if _mpi_disabled():
+        msg = _RAY_NOT_INSTALLED_MSG
+    raise RuntimeError(msg)
```
[the remark this answers is not in the document]
Comment (lori-ren) [reply tensorrt_llm/executor/ray/stub.py:46]: This PR conducts merely a simple move from tensorrt_llm/ray_stub.py to tensorrt_llm/executor/ray/stub.py. I think we should not mix code changes in this PR. Your opinion is right, but this should be done in a follow-up PR, not ours.

File: tensorrt_llm/ray_stub.py [hunk 2]
```diff
@@ -12,35 +12,59 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-from functools import wraps as _wraps
+"""Compatibility shim for ``tensorrt_llm.ray_stub``.
 
-from tensorrt_llm._utils import mpi_disabled as _mpi_disabled
+The implementation moved to ``tensorrt_llm.executor.ray.stub`` in T03
```
Comment (tongyuantongyu) [inline tensorrt_llm/ray_stub.py]: Please rework the comments in this PR (not only this file). Currently they are more like private notes the AI wrote during the refactor, rather than a docstring for future developers to understand the purpose of the codebase and usage of the APIs.

Comment (lori-ren) [reply tensorrt_llm/ray_stub.py]: These comments only appear in shim files which would be eventually removed when T25 is done. The comment follows a previous PR #17254. These shim files are no longer used and only kept for temporary compatibility.

Comment (YihuiLu512) [reply tensorrt_llm/ray_stub.py]: The current PR is a subtask under the code layout Epic (TRTLLM-14558). This stub was introduced as a temporary transitional file to ensure that potential external users reference the correct path. It will be removed in a later release after the entire Epic is completed (expected in 2–3 months), and it will not introduce any changes to the API or the way it is referenced.

The comments in this document already explain the reason for the change and the current editing guidelines for the stub file. Therefore, I believe the current structure can remain as it is.

Comment (tongyuantongyu) [reply tensorrt_llm/ray_stub.py]: 2-3 months are quite long, and there's really no need to put so much detail in the source:

```
Compatibility shim for ``tensorrt_llm.ray_stub``.

Will be removed once all usages are migrated to ``tensorrt_llm.executor.ray.stub``.

DO NOT ADD ANYTHING TO THIS FILE.
```

is probably enough. Can you ask your AI to keep the rest in separated private file and don't commit it?

Comment (lori-ren) [reply tensorrt_llm/ray_stub.py]: Agreed. I could perform a force-push to skip the commit. How about @YihuiLu512?

Comment (YihuiLu512) [reply tensorrt_llm/ray_stub.py]: OK. I’ll open a separate PR to update the shim that has already been merged into the main branch.

Comment (lori-ren) [reply tensorrt_llm/ray_stub.py]: Done. Now all comments of these shim files are trimmed and old ones are skipped. Please check @tongyuantongyu

File: tensorrt_llm/_torch/distributed/communicator.py [hunk 3]
```diff
@@ -30,7 +30,7 @@
 try:
     import ray
 except ModuleNotFoundError:
-    from tensorrt_llm import ray_stub as ray
+    from tensorrt_llm.executor.ray import stub as ray
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/distributed/communicator.py:33]: This is the module that makes the import-weight change matter. Previously the fallback was `from tensorrt_llm import ray_stub as ray` — a leaf module with no `tensorrt_llm` imports. Now it goes through `tensorrt_llm.executor`, whose `__init__.py` eagerly imports `executor`, `proxy`, `worker`, and `result`; that chain transitively reaches this file, so the import is re-entrant against a half-initialized `tensorrt_llm.executor`.

CPython resolves `from pkg import submodule` via `sys.modules` when the parent is mid-init, so this most likely works — but it's the path every install without Ray takes, and it's untested. Please confirm with `python -c "import tensorrt_llm._torch.distributed.communicator"` in a Ray-free env and add it as a CI smoke check.

File: tensorrt_llm/_ray_utils.py [hunk 4]
```diff
@@ -12,47 +12,29 @@
...
+from tensorrt_llm.executor.ray.utils import (  # noqa: F401
+    control_action_decorator,
+    unwrap_ray_errors,
+)
 
-def control_action_decorator(func: Optional[Callable] = None,
-                             *,
-                             drain: bool = True) -> Callable:
-    """Wrap a method in the ``control_action`` context manager.
+warnings.warn(
```
Comment (brnguyen2) [inline tensorrt_llm/_ray_utils.py:30]: `DeprecationWarning` is filtered out by Python's default filters unless it originates in `__main__` or `-W`/`PYTHONWARNINGS` is set — so downstream users (verl, NeMo-RL), the audience these shims exist for, will never see any of these four messages. Use `FutureWarning` if the intent is an actual user-visible nudge. Same applies to `[ray_stub.py:37](https://github.com/NVIDIA/TensorRT-LLM/pull/17487/files#diff-e2a02a03d442c1a60dbb244209dbeee535e13b6e35967ed5ca786432e43be83fR37)`, `[executor/ray_executor.py:27](https://github.com/NVIDIA/TensorRT-LLM/pull/17487/files#diff-3d967a26124faa498e93ac79ec78771a4ae11d7d162e64cd872937250f069f83R27)`, and `[executor/ray_gpu_worker.py:27](https://github.com/NVIDIA/TensorRT-LLM/pull/17487/files#diff-4fd6e239e3b5b71dbe314262b54937c1ddc6bc0ac4ea5951cd4cc7ff421bc6d0R27)`.

File: tensorrt_llm/_ray_utils.py [hunk 5]
```diff
@@ -1,4 +1,4 @@
-# SPDX-FileCopyrightText: Copyright (c) 2022-2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (brnguyen2) [inline tensorrt_llm/_ray_utils.py:1]: The header went from `2022-2024` to `2026`, dropping the original start year on a pre-existing file. Convention for a modified file is to extend the range (`2022-2026`) — which is what `executor/ray/stub.py` does. Worth making the four touched headers consistent.

File: tensorrt_llm/llmapi/rlhf_utils.py [hunk 6]
```diff
@@ -8,9 +8,9 @@
 
 import torch
 
-from tensorrt_llm._ray_utils import control_action_decorator
 from tensorrt_llm._torch.modules.fused_moe.moe_load_balancer import MoeLoadBalancer
 from tensorrt_llm._torch.utils import get_device_uuid
+from tensorrt_llm.executor.ray.utils import control_action_decorator
```
Comment (brnguyen2) [inline tensorrt_llm/llmapi/rlhf_utils.py:13]: This module previously got `control_action_decorator` from a dependency-free leaf; it now imports through the `tensorrt_llm.executor` package init. Since `rlhf_utils` is documented as a `ray_worker_extension_cls` target loaded by qualname inside the worker, worth confirming the added import cost/ordering doesn't change worker startup — a lazy import inside the decorator use site would sidestep it entirely if it does.

File: tensorrt_llm/_ray_utils.py (hunk 5 above)
Comment (lori-ren) [reply tensorrt_llm/_ray_utils.py:1]: `_ray_utils` is not a modified file, but a newly added shim file that serves for compatibility. The original file, which has been moved to executor/ray/utils.py keeps the correct range (2022-2026). So I tend to keep as it is.

File: tensorrt_llm/llmapi/rlhf_utils.py (hunk 6 above)
Comment (lori-ren) [reply tensorrt_llm/llmapi/rlhf_utils.py:13]: `rlhf_utils` already imports from `tensorrt_llm.llmapi`, which further imports from `..executor`. Therefore, before worker loads the target by name, the module cache remains identical as before.

File: tensorrt_llm/_ray_utils.py (hunk 4 above)
Comment (lori-ren) [reply tensorrt_llm/_ray_utils.py:30]: Good catch, will replace these warnings.

File: tensorrt_llm/_torch/distributed/communicator.py (hunk 3 above)
Comment (lori-ren) [reply tensorrt_llm/_torch/distributed/communicator.py:33]: Done. The behavior is now covered by a new `ray_stub_test`, which exercises both paths w or w/o ray installed
